packages/portal-contracts/contracts/messaging/gateway/FuelERC721Gateway.sol

```diff
@@ -0,0 +1,202 @@
+    /// @param to Fuel account to deposit tokens to
+    /// @param tokenAddress Address of the token being transferred to Fuel
+    /// @param fuelContractId ID of the contract on Fuel that manages the deposited tokens
+    /// @param tokenId tokenId to deposit
+    /// @dev Made payable to reduce gas costs
+    function deposit(
+        bytes32 to,
+        address tokenAddress,
+        bytes32 fuelContractId,
+        uint256 tokenId
+    ) external payable whenNotPaused {
+        bytes memory messageData = abi.encodePacked(
+            fuelContractId,
+            bytes32(uint256(uint160(tokenAddress))), // OFFSET_TOKEN_ADDRESS = 32
+            tokenId,
+            bytes32(uint256(uint160(msg.sender))), //from, OFFSET_FROM = 96
+            to, // OFFSET_TO = 128
+            uint256(1) // OFFSET_AMOUNT = 160
+        );
+        _deposit(tokenAddress, fuelContractId, tokenId, messageData);
+    }
+
+    /// @notice Deposits the given tokens to a contract on Fuel with optional data
+    /// @param to Fuel account or contract to deposit tokens to
+    /// @param tokenAddress Address of the token being transferred to Fuel
+    /// @param fuelContractId ID of the contract on Fuel that manages the deposited tokens
+    /// @param tokenId tokenId to deposit
+    /// @param data Optional data to send with the deposit
+    /// @dev Made payable to reduce gas costs
+    function depositWithData(
+        bytes32 to,
+        address tokenAddress,
+        bytes32 fuelContractId,
+        uint256 tokenId,
+        bytes calldata data
+    ) external payable whenNotPaused {
+        bytes memory messageData = abi.encodePacked(
+            fuelContractId,
+            bytes32(uint256(uint160(tokenAddress))), // OFFSET_TOKEN_ADDRESS = 32
+            tokenId, // OFFSET_TOKEN_ID = 64
+            bytes32(uint256(uint160(msg.sender))), //from, OFFSET_FROM = 96
+            to, // OFFSET_TO = 128
+            uint256(1), // OFFSET_AMOUNT = 160
+            DEPOSIT_TO_CONTRACT, // OFFSET_ROLE = 161
+            data
+        );
+        _deposit(tokenAddress, fuelContractId, tokenId, messageData);
+    }
+
+    /// @notice Finalizes the withdrawal process from the Fuel side gateway contract
+    /// @param to Account to send withdrawn tokens to
+    /// @param tokenAddress Address of the token being withdrawn from Fuel
+    /// @param tokenId Discriminator for ERC721 / ERC1155 tokens
+    /// @dev Made payable to reduce gas costs.
+    /// @dev Could remove the amount param to further reduce cost, but that implies changes in the Fuel contract
+    function finalizeWithdrawal(
+        address to,
+        address tokenAddress,
+        uint256 /*amount*/,
+        uint256 tokenId
+    ) external payable override whenNotPaused onlyFromPortal {
+        bytes32 fuelContractId = messageSender();
+        require(_deposits[tokenAddress][tokenId] == fuelContractId, "Fuel bridge does not own this token");
+
+        delete _deposits[tokenAddress][tokenId];
```

Comment:
Is it cheaper to `delete` or override with a default value? Is it safe to use some sentinel value as a default?

Reply:
Depends on the context:
- Deleting (or setting from non-0 to 0 value, it is the same action) has the effect of releasing storage space for the blockchain, which is rewarded with decreased gas costs - ie. the specific withdrawal transaction would be cheaper. WRT the use of `delete` instead of assigning a null value, it is purely opinionated - I tend to use `delete` because I feel it is more idiomatic.
-  Setting a non-0 value to another non-0 (sentinel) value is cheaper than going from 0 to non-0. This would make withdrawal a bit more expensive, and deposit a bit cheaper _when the asset has already been bridged before_. Also we 'd need to check for both the null value and the sentinel value, because we cannot initialise storage to a default / sentinel value.

Overall I think the sentinel value would be a bit more expensive long term.